pkg/services/annotations/annotationsimpl/annotations.go

```diff
@@ -72,12 +72,45 @@ func (r *RepositoryImpl) Update(ctx context.Context, item *annotations.Item) err
 }
 
 func (r *RepositoryImpl) Find(ctx context.Context, query *annotations.ItemQuery) ([]*annotations.ItemDTO, error) {
-	resources, err := r.authZ.Authorize(ctx, query.OrgID, query)
-	if err != nil {
-		return make([]*annotations.ItemDTO, 0), err
+	if query.Limit == 0 {
+		query.Limit = 1000
 	}
 
-	return r.reader.Get(ctx, query, resources)
+	// Return early if no annotations found, it's not necessary to perform expensive access control filtering
+	res, err := r.reader.Get(ctx, query, &accesscontrol.AccessResources{
+		Dashboards:               map[string]int64{},
+		CanAccessDashAnnotations: true,
+		CanAccessOrgAnnotations:  true,
+	})
+	if err != nil || len(res) == 0 {
+		return []*annotations.ItemDTO{}, err
+	}
+
+	results := make([]*annotations.ItemDTO, 0, query.Limit)
+	query.Page = 1
+
+	// Iterate over available annotations util query limit is reached
+	// or all available dashboards are checked
+	for len(results) < int(query.Limit) {
+		resources, err := r.authZ.Authorize(ctx, query)
+		if err != nil {
+			return []*annotations.ItemDTO{}, err
```

Comment:
I also typically just return `nil` for a slice or map value on a return in the error case.